test/subscription_map.h

```diff
@@ -0,0 +1,265 @@
+//
+// Created by wkl04 on 10-12-2019.
+//
+
+#ifndef MQTTSUBSCRIPTION_SUBSCRIPTION_MAP_H
+#define MQTTSUBSCRIPTION_SUBSCRIPTION_MAP_H
+
+#include <mqtt/string_view.hpp>
+
+#include <unordered_map>
+#include <boost/functional/hash.hpp>
+
+#include "path_tokenizer.h"
+
+/**
+ *
+ * Non-normative comment
+        · $SYS/ has been widely adopted as a prefix to topics that contain Server-specific information or control APIs
+        · Applications cannot use a topic with a leading $ character for their own purposes
+
+    Non-normative comment
+        · A subscription to “#” will not receive any messages published to a topic beginning with a $
+        · A subscription to “+/monitor/Clients” will not receive any messages published to “$SYS/monitor/Clients”
+        · A subscription to “$SYS/#” will receive messages published to topics beginning with “$SYS/”
+        · A subscription to “$SYS/monitor/+” will receive messages published to “$SYS/monitor/Clients”
+        · For a Client to receive messages from topics that begin with $SYS/ and from topics that don’t begin with a $, it has to subscribe to both “#” and “$SYS/#”
+ */
+
+template<typename Value>
+class subscription_map_base
+{
+    using node_id = size_t;
+    using path_entry_key = std::pair< node_id, std::string>;
+
+    static constexpr node_id root_node_id = 0;
+
+    struct path_entry
+    {
+        node_id id;
+
+        bool has_hash_child : 1;
+        bool has_plus_child : 1;
+        uint32_t count : 30;
```

Comment:
on a 64bit platform, `node_id` is 64bits, meaning that for alignment reasons the compiler is going to insert an additional 32 bits to the end of your struct here.

Might as well grow the struct to take advantage of that extra space.

Reply:
How? And count would become hardware dependent because of this ..

Reply:
> How?

Not sure what you're asking.

>  And count would become hardware dependent because of this ..

Don't see a problem with that. Better hardware -> better functionality.